econet: Add EN7528 subtarget and DASAN H660GM-A device

naseef opened the pull request:
This PR adds support for the EN7528 subtarget and the first device, DASAN H660GM-A.

  The EN7528 is a little endian dual-core MIPS 1004Kc SoC used in xPON devices. Unlike the big endian EN751221 family, EN7528 uses the MIPS GIC interrupt controller for SMP and requires little endian TRX firmware headers.

  The DASAN H660GM-A is a GPON ONT with 256MB RAM, 256MB SPI NAND flash, 4x 1GbE ethernet ports, 1 FXS port, 1 GPON port, and 1 USB 2.0 port.

  Changes included:

  1. EN7528 subtarget support - New mipsel subtarget with kernel patches for GIC interrupt controller, custom timer driver using GIC shared interrupts per VPE, and SPI driver fix for EN7528 chip select handling.
  2. Little endian TRX support - The tclinux TRX firmware format stores fields in the CPU's native byte order. Updated tclinux-trx.sh to support both big endian (EN751221, magic "2RDH") and little endian (EN7528, magic "HDR2") variants.
  3. Configurable BBT table size - Different vendor firmware versions use different BBT table sizes. Added DTS property econet,bbt-table-size to configure the table size, fixing BBT checksum validation failures on devices with non-default table sizes.
  4. DASAN H660GM-A device support - Device tree, image configuration, and dual-boot support via en75_chboot.
  5. Basic Ehthenet support
  6. GPIO Support
  7. PCIe/WiFi support

  Flashing instructions:

  The DASAN H660GM-A can be flashed directly from the OEM firmware web UI:
  1. Login with username superuser and password Dz$!A!r7.
  2. Rename openwrt-econet-en7528-dasan_h660gm-a-squashfs-tclinux.trx to G_ONU_openwrt.bin.
  3. Upload via Maintenance > Firmware Upgrade.

  Other EN7528 devices can either be booted with initramfs or flashed using the bootloader flash command, same as EN751221 devices since they share a similar bootloader.

  Like EN751221, the EN7528 devices use a dual-image layout. The design preserves factory firmware data and supports dual-boot between OpenWrt and the factory OS using en75_chboot.

naseef committed:
econet: en75_bmt: add configurable BBT table size

Different vendor firmware versions use different BBT table sizes. The
checksum is calculated over the entire table, so the size must match
what the bootloader expects.

The Genexis Platinum-4410 bootloader was compiled with a BBT table
size of 250 entries (MAX_RAW_BAD_BLOCK_SIZE as found in vendor code).
Without this fix, the BBT checksum validation fails:

    [    0.391948] spi-nand spi0.0: Dosilicon SPI NAND was found.
    [    0.397651] spi-nand spi0.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
    [    0.407370] en75_bmt: found BMT in block 1023
    [    0.450160] en75_bmt: BBT not found and econet,can-write-factory-bbt is unset, giving up

Add a new DTS property 'econet,bbt-table-size' to configure the BBT
table size. If not specified, defaults to 1000.

After this patch:

    [    0.407021] en75_bmt: found BMT in block 1023
    [    0.449159] en75_bmt: found BBT in block 943
    [    0.453491] en75_bmt: BBT & BMT found
    [    0.457152] en75_bmt: blocks: total: 1024, user: 943, factory_bad: 0, worn: 0 reserve: 81
    [    0.465390] en75_bmt: 117 MiB usable space

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef committed:
econet: add EN7528 subtarget support

The EN7528 is a little endian dual-core MIPS 1004Kc SoC used in xPON
devices. Unlike the big endian EN751221, EN7528 uses the MIPS GIC
interrupt controller for SMP.

This adds minimal boot support for EN7528:
- New en7528 subtarget with mipsel architecture
- Kernel patches for EN7528 SoC with GIC support
- Timer driver extended to support GIC shared interrupts per CPU
- SPI driver fix for EN7528 chip select handling
- Generic device tree for initial bring-up

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef committed:
econet: image: add little endian TRX support for EN7528

The EN7528 SoC uses a little endian MIPS architecture, unlike the big
endian EN751221 family. The tclinux TRX firmware format stores multi-byte
fields in the CPU's native byte order, requiring different header layouts
for each architecture:

  - Big endian (EN751221): magic "2RDH", fields in big endian order
  - Little endian (EN7528): magic "HDR2", fields in little endian order

Update tclinux-trx.sh to support both endianness variants:
  - Add --endian parameter to select byte order (default: be)
  - Add --model parameter for optional platform identifier field
  - Convert to named parameters for clarity and extensibility
  - Use hex32() helper for endian-aware 32-bit field output

Move TRX_ENDIAN configuration to subtarget files, allowing each subtarget
to specify its native byte order:
  - en751221.mk: TRX_ENDIAN := be
  - en7528.mk: TRX_ENDIAN := le

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef committed:
econet: en7528: add support for DASAN H660GM-A

DASAN H660GM-A is a GPON ONT.

Specification:
- SoC:      EcoNet EN7528
- RAM:      256 MB
- Flash:    256 MB SPI NAND
- Ethernet: 4x 1GbE
- FXS:      1 port
- GPON:     1 port
- USB:      1x USB 2.0

Install via OEM web UI:
1. Login to the factory web UI with username "superuser"
   and password "Dz$!A!r7".
2. Rename the OpenWrt image
   openwrt-econet-en7528-dasan_h660gm-a-squashfs-tclinux.trx
   to G_ONU_openwrt.bin.
3. Upload the renamed image via the firmware upgrade page
   at Maintenance > Firmware Upgrade in the factory web UI.

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef commented:
<details>

<summary>Here is openwrt bootlog</summary>

```

Uncompress kernel to 0x80020000 (free_mem_ptr=0x80e00000 free_mem_ptr_end=0x807b0000)
Uncompressing [LZMA] ...  done.
booting kernel from 0x80020000...
zimage at:     80023390 802F8B60
Uncompressing Linux at load address 81000000
Copy device tree to address  8196D720
Now, booting the kernel...
[    0.000000] Linux version 6.12.62 (user@host) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 14.3.0 r32284+8-4e4986aaf6) 14.3.0, GNU ld (GNU Binutils) 2.44) #0 SMP Tue Dec 30 06:08:45 2025
[    0.000000] printk: legacy bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001992f (MIPS 1004Kc)
[    0.000000] MIPS: machine is DASAN H660GM-A
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] OF: reserved mem: Reserved memory: No reserved-memory node in the DT
[    0.000000] VPE topology {2,2} total 4
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] percpu: Embedded 12 pages/cpu s18192 r8192 d22768 u49152
[    0.000000] Kernel command line:
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Writing ErrCtl register=0007f3d0
[    0.000000] Readback ErrCtl register=0007f3d0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 65536
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] RCU Tasks Trace: Setting shift to 2 and lim to 1 rcu_task_cb_adjust=1 rcu_task_cpu_ids=4.
[    0.000000] NR_IRQS: 256
[    0.000000] rcu: srcu_init: Setting srcu_struct sizes based on contention.
[    0.000000] clocksource: timer: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 9556302233 ns
[    0.000005] sched_clock: 32 bits at 200MHz, resolution 5ns, wraps every 10737418237ns
[    0.008531] timer: using 200.000 MHz high precision timer
[    0.014690] Calibrating delay loop... 597.60 BogoMIPS (lpj=2988032)
[    0.081385] pid_max: default: 32768 minimum: 301
[    0.096383] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.104365] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.122903] rcu: Hierarchical SRCU implementation.
[    0.128183] rcu:     Max phase no-delay instances is 1000.
[    0.135037] smp: Bringing up secondary CPUs ...
[    0.141011] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.141048] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.141064] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.141114] CPU1 revision is: 0001992f (MIPS 1004Kc)
[    0.202151] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.202178] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.202192] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.202224] CPU2 revision is: 0001992f (MIPS 1004Kc)
[    0.262498] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.262527] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.262542] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.262578] CPU3 revision is: 0001992f (MIPS 1004Kc)
[    0.321571] smp: Brought up 1 node, 4 CPUs
[    0.410360] Memory: 246984K/262144K available (7386K kernel code, 629K rwdata, 1432K rodata, 1232K init, 227K bss, 13644K reserved, 0K cma-reserved)
[    0.427293] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.438139] futex hash table entries: 1024 (order: 3, 32768 bytes, linear)
[    0.454307] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.472181] clocksource: Switched to clocksource timer
[    0.487723] NET: Registered PF_INET protocol family
[    0.493300] IP idents hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.502094] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.511298] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.519838] TCP established hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.528243] TCP bind hash table entries: 2048 (order: 3, 32768 bytes, linear)
[    0.536193] TCP: Hash tables configured (established 2048 bind 2048)
[    0.544139] MPTCP token hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.552444] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.559574] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.568214] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.577847] workingset: timestamp_bits=14 max_order=16 bucket_order=2
[    0.587731] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.596519] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.605565] printk: legacy console [ttyS0] disabled
[    0.611770] 1fbf0000.serial: ttyS0 at MMIO 0x1fbf0000 (irq = 20, base_baud = 115200) is a 16550
[    0.621412] printk: legacy console [ttyS0] enabled
[    0.621412] printk: legacy console [ttyS0] enabled
[    0.631424] printk: legacy bootconsole [early0] disabled
[    0.631424] printk: legacy bootconsole [early0] disabled
[    0.649220] spi-nand spi0.0: Micron SPI NAND was found.
[    0.654544] spi-nand spi0.0: 256 MiB, block size: 128 KiB, page size: 2048, OOB size: 128
[    0.664843] en75_bmt: found BMT in block 2047
[    0.952935] en75_bmt: found BBT in block 1885
[    0.957335] en75_bmt: BBT & BMT found
[    0.961005] en75_bmt: blocks: total: 2048, user: 1885, factory_bad: 0, worn: 0 reserve: 163
[    0.969391] en75_bmt: 235 MiB usable space
[    0.970007] 9 fixed-partitions partitions found on MTD device spi0.0
[    0.980531] Creating 9 MTD partitions on "spi0.0":
[    0.985355] 0x000000000000-0x000000040000 : "bootloader"
[    0.992830] 0x000000040000-0x000000080000 : "dzs"
[    0.998986] 0x000000080000-0x000002080000 : "tclinux"
[    1.056706] 0x000000480000-0x000002080000 : "rootfs"
[    1.062977] mtd: setting mtd3 (rootfs) as root device
[    1.069798] 1 squashfs-split partitions found on MTD device rootfs
[    1.076037] 0x000000700000-0x000002080000 : "rootfs_data"
[    1.082667] 0x000002080000-0x000004080000 : "tclinux_slave"
[    1.141125] 0x000004080000-0x0000040c0000 : "romfile"
[    1.147628] 0x0000040c0000-0x0000060c0000 : "rwfs"
[    1.205484] 0x0000060c0000-0x00000ddc0000 : "unknown"
[    1.414041] 0x00000ddc0000-0x00000e000000 : "reservearea"
[    1.433384] NET: Registered PF_INET6 protocol family
[    1.441440] Segment Routing with IPv6
[    1.445299] In-situ OAM (IOAM) with IPv6
[    1.449348] NET: Registered PF_PACKET protocol family
[    1.454524] 8021q: 802.1Q VLAN Support v1.8
[    1.484004] clk: Disabling unused clocks
[    1.488609] mtdblock: MTD device 'rootfs' is NAND, please consider using UBI block devices instead.
[    1.530411] VFS: Mounted root (squashfs filesystem) readonly on device 31:3.
[    1.541822] Freeing unused kernel image (initmem) memory: 1232K
[    1.547989] This architecture does not have kernel memory protection.
[    1.554516] Run /sbin/init as init process
[    3.773579] init: Console is alive
[    4.993536] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    5.178904] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    5.189962] init: - preinit -
[    9.132199] random: crng init done
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[   13.530809] mount_root: no usable overlay filesystem found, using tmpfs overlay
[   13.542559] urandom-seed: Seed file not found (/etc/urandom.seed)
[   13.701667] procd: - early -
[   14.438101] procd: - ubus -
[   14.850566] procd: - init -
Please press Enter to activate this console.
[   15.893525] kmodloader: loading kernel modules from /etc/modules.d/*
[   16.143492] PPP generic driver version 2.4.2
[   16.149624] NET: Registered PF_PPPOX protocol family
[   16.161315] kmodloader: done loading kernel modules from /etc/modules.d/*
[   16.892540] urngd: v1.0.2 started.


BusyBox v1.37.0 (2025-12-16 08:05:12 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r32420+3-d9b1c74780
 -----------------------------------------------------

 === WARNING! =====================================
 There is no root password defined on this device!
 Use the "passwd" command to set up a new password
 in order to prevent unauthorized SSH logins.
 --------------------------------------------------


 OpenWrt recently switched to the "apk" package manager!

 OPKG Command           APK Equivalent      Description
 ------------------------------------------------------------------
 opkg install <pkg>     apk add <pkg>       Install a package
 opkg remove <pkg>      apk del <pkg>       Remove a package
 opkg upgrade           apk upgrade         Upgrade all packages
 opkg files <pkg>       apk info -L <pkg>   List package contents
 opkg list-installed    apk info            List installed packages
 opkg update            apk update          Update package lists
 opkg search <pkg>      apk search <pkg>    Search for packages
 ------------------------------------------------------------------

For more information visit:
https://openwrt.org/docs/guide-user/additional-software/opkg-to-apk-cheatsheet

root@OpenWrt:~#

```
<details>

abajk commented:
Can you adapt the [upstream driver](https://github.com/openwrt/openwrt/pull/20365#issuecomment-3391261041) to support SPI on EN7581?

cjdelisle commented on target/linux/econet/patches-6.12/101-econet-timer-add-en7528-support.patch:
```diff
@@ -0,0 +1,289 @@
++/*
++ * Timer present on EcoNet EN7528 MIPS based SoCs.
++ *
++ * Unlike EN751221 which uses a single percpu IRQ, EN7528 uses separate
```
So you suppose there is a way to make the timer able to do both modes? I assume everything with GIC is going to be like this, so that includes also the EN751627, and "there's no code like no code" so one timer is better than two.

naseef replied:
let me try. What do you suggest? as part of this PR or separate? Also one question, for this PR to be merged, is it needed to be in upstream?

abajk commented on target/linux/econet/dts/en7528.dtsi:
```diff
@@ -0,0 +1,97 @@
+		interrupts = <GIC_SHARED 2 IRQ_TYPE_LEVEL_HIGH>;
+
+		/*
+		 * Conversion of baud rate to clock frequency requires a
```
On airoha, we have [this code](https://github.com/openwrt/openwrt/blob/main/target/linux/airoha/patches-6.12/886-uart-add-en7523-support.patch) used to control the baud rate. Isn't this the same IP Core as on AN7523/AN7581?

naseef committed:
econet: spi: sync spi-airoha-snfi patches from airoha target

Copy upstream v6.19 spi-airoha-snfi driver patches from
target/linux/airoha/patches-6.12 for the EN7528 subtarget.

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef commented on target/linux/econet/patches-6.12/100-econet-add-en7528-soc.patch:
```diff
@@ -0,0 +1,111 @@
++		select SYS_SUPPORTS_MULTITHREADING
++		select SYS_SUPPORTS_SMP
++		help
++		  The EN7528/EN751627 family with dual-core MIPS 1004Kc.
```
done

cjdelisle commented:
> * unified timer drivers. tested on both en751221 and en7528

I *think* we will need to keep the timer called `timer-econet-en751221.c` and symbol `CONFIG_ECONET_EN751221_TIMER` because I suspect the kernel guys are going to read the patch and say "Why are you changing it? Even if it's not the ideal name - what if someone is using it already? When the symbol changes their kernel config will stop working". If you know better than me what is considered best practice, I would be open to your perspective.

cjdelisle commented:
Your changes to the timer look okay to my eyes. If you would be willing to send it upstream, I would appreciate it. If you don't have time, I could do it but I will need you to update the patch so that it has a `Signed-off-by` because the kernel guys are pretty strict about copyright licensing.

naseef commented on target/linux/econet/patches-6.12/100-econet-add-en7528-soc.patch:
```diff
@@ -0,0 +1,107 @@
+-	select CPU_BIG_ENDIAN
+ 	select DEBUG_ZBOOT if DEBUG_KERNEL
+ 	select EARLY_PRINTK_8250
+-	select ECONET_EN751221_TIMER
```
Should I to add back to the main ECONET instead of separately adding to subtargets?

cjdelisle commented:
> Also, should I amend the unified timer commit into the original EN7528 support commit for this PR

However is easiest, I just need a `Signed-off-by` in the patch so I have a history for the patch.

cjdelisle commented:
@naseef FYI I just started an EN7528 build repo: https://github.com/cjdelisle/OpenWrt-EN7528-Builds/releases/latest , if you could give the trx a test, that would be awesome.

naseef committed:
econet: en7528: add basic ethernet support

EN7528 shares the same clock/reset controller as EN7523. Enable
COMMON_CLK_EN7523 and RESET_CONTROLLER for ethernet hardware resets.
Update econet-eth driver and add it as default package.

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef committed:
econet: en7528: add GPIO and LED support for DASAN H660GM-A

Enable the Airoha EN7523 GPIO driver for EN7528 and add GPIO
controller nodes to the EN7528 DTSI. Add LED, button and GPIO
definitions for the DASAN H660GM-A board.

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

naseef committed:
econet: en7528: add PCIe and WiFi support

Add PCIe controller and PHY support for EN7528 SoC. This includes
a new PCIe PHY driver, EN7528-specific startup in the MediaTek PCIe
controller, and a fix for bogus prefetch window reads on bridges
that do not implement the registers.

Enable WiFi for the DASAN H660GM-A board with MT7603 (2.4 GHz) and
MT7615/MT7663 (5 GHz).

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21326
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

EmreYavuzalp commented:
I am also compiling this for my similiar devices. Let's see what happens. Gonna try on initramfs and then edit according to my needs.

makver9 commented:
It has 128MB flash. 

nand: device found, Manufacturer ID: 0x2c, Chip ID: 0xf1
nand: Micron MT29F01G08ABAEA
nand: 128MiB,

Have we not handled that yet ? What should we change to support it ?

dreamer1001 commented:
@AKoo7 is the 4 lan ports working on 414
